notepad-plus-plus/notepad-plus-plus
# FEATURE: Notify plugins of the user's native language preference (would address #15513)

### Rationale

This PR implements `NPPN_RELOADNATIVELANG`, which makes it easy for plugins to discover the user's native language preference without having to parse the `nativeLang.xml` file or subclass the Notepad++ native window, [as illustrated here](https://github.com/molsonkiko/NppCSharpPluginPack/blob/main/NppCSharpPluginPack/Utils/Translator.cs).

Arguments in favor of implementing a notification like this are discussed at length in #15513. Given the relatively low code footprint of this implementation, and the fact that it mostly is just doing work that each plugin wishing to translate itself otherwise would have had to do itself, this seems like a net improvement, assuming we are serious about trying to make Notepad++ as user-friendly as possible for non-English speakers. 

### Description of implementation

The `SCNotification` structs of `NPPN_RELOADNATIVELANG` have the following fields filled:
```
scnNotification->nmhdr.code = NPPN_RELOADNATIVELANG;
scnNotification->nmhdr.hwndFrom = hPluginMenu; // the menu handle to the plugin menu, so that plugins can find their plugin's submenu and re-translate their menu items
scnNotification->nmhdr.idFrom = languageName; // the ANSI encoded English name of the current UI language for Notepad++ (this is the same as the name of the localization XML file used by Notepad++ to translate the UI).
```

__EDIT:__ As this notification is currently implemented, __the `hwndFrom` and `idFrom` *must not be `NULL`* for the notification to be sent.__

This notification is emitted once at startup (at the very end of `Notepad_plus::init`) and also once each time the user changes their language preference in the Preferences form.

In order to avoid having to send this message twice every time the user changes their language preference, this PR has also changed `NPPM_INTERNAL_RELOADNATIVELANG`, changing the previously unused `WPARAM` into a boolean that, if `TRUE`, indicates that plugins should be notified of the native language change.

Aside from new documentation, most of the code footprint of this PR is associated with finding the menu handle of the plugins menu, which most plugins will need so that they can find the menu handle of their plugin's submenu. This code could probably be simplified by assuming that the plugins menu is always at `MENUINDEX_PLUGINS`, but this will not always be true, because some plugins like [Columns++](https://github.com/Coises/ColumnsPlusPlus) add their plugin menu to the Notepad++ main menu, causing the plugins menu to be at a different position than it normally would be.

### How to test

To see an example of how a plugin could respond to this new notification, see [this branch of NppCSharpPluginPack](https://github.com/molsonkiko/NppCSharpPluginPack/tree/nppn_reloadnativelang-proof_of_concept).

You can test this PR by:
1. Build this PR, or download the GitHub artifact and place it in the installation directory of an existing installation of Notepad++ 8.6.9.
2. Make sure the installation directory of this Notepad++ installation has a folder named "localization" with the `spanish.xml` and `english.xml` files downloaded from [the Notepad++ master `nativeLang` folder](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/installer/nativeLang/spanish.xml)
3. Download the appropriate version of my testing fork of NppCSharpPluginPack:
    - 64 bit: https://github.com/molsonkiko/NppCSharpPluginPack/blob/nppn_reloadnativelang-proof_of_concept/NppCSharpPluginPack/Release_x64.zip
    - 32 bit: https://github.com/molsonkiko/NppCSharpPluginPack/blob/nppn_reloadnativelang-proof_of_concept/NppCSharpPluginPack/Release_x86.zip
4. Extract the files from "Release_x64.zip" or "Release_x86.zip" into a folder named "CSharpPluginPack".
5. Copy the "CSharpPluginPack" folder into the "plugins" folder of the installation folder that contains this build of Notepad++.
6. Run Notepad++
7. Go to the `Plugins->CSharpPluginPack` menu, and open up the `Selection Remembering Form`. A docking form will appear on the right side of the Notepad++ window.
8. Go to the `Settings->Preferences` form, and change Localization to "Español".
9. Verify that the controls on the Selection Remembering Form are now in Spanish.
10. Go to the `Complementos->CSharpPluginPack` submenu, and verify that all the menu items on that submenu are in Spanish. Try clicking "Sobre" and verify that it opens a modal dialog that says "Sobre NppCSharpPluginPack".
11. Change the Localization back to "English".
12. Verify that the `Complementos->CSharpPluginPack` menu items and forms are now in English.
13. Change the Localization back to Spanish, then close Notepad++.
14. Reopen Notepad++. Go to `Complementos->CSharpPluginPack`, and verify that everything is in Spanish (steps 7 and 10, except `Selection Remembering Form` has become `Formulario para recordar las selecciónes`).
15. Change the Localization back to English, and verify that everything in CSharpPluginPack is translated back to English.
16. Repeat steps 1-15 for any other builds of this commit that you wish to test.
